master/buildbot/worker/kubernetes.py

```diff
@@ -0,0 +1,368 @@
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Copyright Buildbot Team Members
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from future.utils import PY2
+from future.utils import string_types
+
+import os
+import socket
+import tempfile
+
+from twisted.internet import defer
+
+from buildbot.interfaces import LatentWorkerFailedToSubstantiate
+from buildbot.util.httpclientservice import HTTPClientService
+from buildbot.util.logger import Logger
+from buildbot.worker.docker import DockerBaseWorker
+
+try:
+    import yaml
+    HAS_YAML = True
+except ImportError:
+    HAS_YAML = False
+
+if PY2:
+    FileNotFoundError = IOError
+
+
+log = Logger()
+
+
+class KubeClientService(HTTPClientService):
+
+    default_kubeconfig = '~/.kube/config'
+
+    kube_dir = '/var/run/secrets/kubernetes.io/serviceaccount/'
+
+    kube_namespace_file = os.path.join(kube_dir, 'namespace')
+    kube_token_file = os.path.join(kube_dir, 'token')
+    kube_cert_file = os.path.join(kube_dir, 'ca.crt')
+
+    config_loader_list = ['kube_config_loader', 'incluster_config_loader']
+
+    def __init__(self,
+                 kubeconfig=None,
+                 config_method=None):
+        self.kube_config_file = kubeconfig or self.default_kubeconfig
+        self._cert_file = None
+        url, req_kwargs, namespace = self.guess_config(config_method)
+        self._cert = req_kwargs.pop('cert', None)
+        self._verify = req_kwargs.pop('verify', None)
+        HTTPClientService.__init__(self, url, **req_kwargs)
+        self.namespace = namespace
+
+    def guess_config(self, config_method=None):
+        if isinstance(config_method, string_types):
+            config_loader_iter = iter((config_method,))
+        elif config_method is None:
+            config_loader_iter = iter(self.config_loader_list)
+        else:
+            try:
+                config_loader_iter = iter(config_method)
+            except TypeError:
+                config_loader_iter = iter((config_method,))
+        for config in config_loader_iter:
+            if callable(config):
+                config_loader = config
+            else:
+                try:
+                    config_loader = getattr(self, config)
+                except AttributeError:
+                    raise ValueError(
+                        "Unknown config method '{0}'. "
+                        "Only {1} are available.".format(
+                            config, self.config_loader_list
+                        )
+                    )
+            if not config_loader:
+                continue
+            try:
+                return config_loader()
+            except KubeClientConfigException:
+                continue
```

Comment:
what if there is a syntax error in my config? this code won't allow to debug that.